miio/gateway.py

```diff
@@ -126,58 +125,64 @@ def light(self) -> "GatewayLight":
         """Return light control interface."""
         return self._light
 
-    @command()
+    @property
     def devices(self):
-        """Return list of devices."""
+        """Return a list of the already discovered devices."""
+        return self._devices
+
+    @command()
+    def discover_devices(self):
+        """Discovers SubDevices and returns a list of the discovered devices."""
         # from https://github.com/aholstenson/miio/issues/26
-        devices_raw = self.send("get_device_prop", ["lumi.0", "device_list"])
-        devices = [
-            SubDevice(self, *devices_raw[x : x + 5])  # noqa: E203
-            for x in range(0, len(devices_raw), 5)
-        ]
+        device_type_mapping = {
+            "AqaraRelayTwoChannels": AqaraRelayTwoChannels,
+            "Plug": AqaraPlug,
+            "SensorHT": SensorHT,
+            "AqaraHT": AqaraHT,
+            "AqaraMagnet": AqaraMagnet,
+        }
+        devices_raw = self.get_gateway_prop("device_list")
+        self._devices = []
+
+        for x in range(0, len(devices_raw), 5):
+            try:
+                device_name = DeviceType(devices_raw[x + 1]).name
+            except ValueError:
+                _LOGGER.warning(
+                    "Unknown subdevice type '%i': %s discovered, of Xiaomi gateway with ip: %s",
+                    devices_raw[x + 1],
+                    devices_raw[x],
+                    self.ip,
+                )
+                device_name = DeviceType(-1).name
+
+            subdevice_cls = device_type_mapping.get(device_name)
+            if subdevice_cls is None and device_name != DeviceType(0).name:
```

Comment:
So using enum and skipping the name checking, the complete sub-device initialization logic would be following:
1. Construct DeviceType
2. If it fails (because of missing type, print out the warning like right now)
3. Check `if dev_type == DeviceType.Gateway`, ignore gateways (just `continue` the loop)
4. Initialize the device class.

Reply:
it is nice to check if the dev_type (int) is in the enum and print a warning if it is not, this allows us to discover more dev_types from users reporting those warnings.

Reply:
We can still print the value when the conversion fails, and ask the users to create an issue :-)